Python/Product/Analysis/ModuleResolver.cs

```diff
@@ -0,0 +1,206 @@
+        /// </param>
+        /// <returns>
+        /// A sequence of strings representing the absolute names of the module
+        /// in order of precedence.
+        /// </returns>
+        internal static IEnumerable<string> ResolvePotentialModuleNames(
+            string importingFromModuleName,
+            string importingFromFilePath,
+            string relativeModuleName,
+            bool absoluteImports
+        ) {
+            string importingFrom = null;
+            if (!string.IsNullOrEmpty(importingFromModuleName)) {
+                importingFrom = importingFromModuleName;
+                if (!string.IsNullOrEmpty(importingFromFilePath) && ModulePath.IsInitPyFile(importingFromFilePath)) {
+                    if (string.IsNullOrEmpty(importingFrom)) {
+                        importingFrom = "__init__";
+                    } else {
+                        importingFrom += ".__init__";
+                    }
+                }
+            }
+
+            if (string.IsNullOrEmpty(relativeModuleName)) {
+                yield break;
+            }
+
+            // Handle relative module names
+            if (relativeModuleName.FirstOrDefault() == '.') {
+                yield return GetModuleFullName(importingFrom, relativeModuleName);
+                yield break;
+            }
+
+            // The two possible names that can be imported here are:
+            // * relativeModuleName
+            // * importingFrom.relativeModuleName
+            // and the order they are returned depends on whether
+            // absolute_import is enabled or not.
+
+            // Assume trailing dots are not part of the import
+            relativeModuleName = relativeModuleName.TrimEnd('.');
+
+            // With absolute_import, we treat the name as complete first.
+            if (absoluteImports) {
+                yield return relativeModuleName;
+            }
+
+            if (!string.IsNullOrEmpty(importingFrom)) {
+                var prefix = importingFrom.Split('.');
+
+                if (prefix.Length > 1) {
+                    var adjacentModuleName = string.Join(".", prefix.Take(prefix.Length - 1)) + "." + relativeModuleName;
+                    yield return adjacentModuleName;
+                }
+            }
+
+            // Without absolute_import, we treat the name as complete last.
+            if (!absoluteImports) {
+                yield return relativeModuleName;
+            }
+        }
+
+        private static string GetModuleFullName(string originatingModule, string relativePath) {
+            // Check if it is indeed relative
+            if (string.IsNullOrEmpty(relativePath) || relativePath[0] != '.') {
+                return relativePath;
+            }
+
+            var up = relativePath.TakeWhile(ch => ch == '.').Count();
+            var bits = originatingModule.Split('.');
+            if (up > bits.Length) {
+                return relativePath; // too far up
```

Comment:
We should probably return `null` here to indicate that there's nothing to import